docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,177 @@
+  at `/apis/<groupVersion>/`)
+* No bootstrap problem: The core kubernetes server should not depend on any
+  other federated server to come up. Other servers can only depend on the core
+  kubernetes server.
+
+## Implementation Details
+
+### Reverse proxy
+
+A very basic Go program will suffice as a reverse proxy. Each federated
+API server will register its baseURL and swagger spec path with the proxy. This reverse proxy
+will summarize the list of all group versions exposed by all registered API
+servers at `/apis`.
+The Go program can also be replaced by any of the multiple open source API management tools
+(for example, there is [Kong](https://getkong.org/), which is written in lua and there is
+[Tyk](https://tyk.io/), which is written in Go).
+
+### Storage
+
+Each API server is responsible for storing their resources. They can have their
+own etcd or can use kubernetes server's etcd using [third party
+resources](../design/extending-api.md#adding-custom-resources-to-the-kubernetes-api-server).
+
+### Health check
+
+Kubernetes server's `/api/v1/componentstatuses` will continue to report status
+of master components that it depends on (scheduler and various controllers).
+Similiarly, proxy will expose a health check endpoint that will report the status of all
+federated api servers in the cluster.
+
+### Auth
+
+Since, the actual server which serves client's request is opaque to the client,
+all API servers need to have homogeneous authentication and authorisation mechanisms.
+All API servers will handle authn and authz for their resources themselves,
+with proxy being a simple pass through.
+Future proposals will refine how auth is managed across the cluster.
+
+### kubectl
+
+kubectl will talk to the proxy and use the discovery API to figure out the
+operations and resources supported in the cluster.
+Today, it uses RESTMapper to determine that. We will update kubectl code to populate
+RESTMapper using the discovery API so that we can add and remove resources
+at runtime.
+We will also need to make kubectl truly generic. Right now, a lot of operations
+(like get, describe) are hardcoded in the binary for all resources. A future
+proposal will provide details on moving those operations to server.
+
+### Handling global policies
+
+Now that we have resources spread across multiple API servers, we need to
+be careful to ensure that global policies are enforced.
+Future proposals will improve how this is done across the cluster.
+
+#### Namespaces
+
+When a namespaced resource is created in any of the federated server, that
+server first needs to check with the kubernetes server that:
+
+* The namespace exists
+* User has authorization to create resources in that namespace
+* Resource quota for the namespace is not exceeded.
+
+While deleting a namespace, kubernetes server needs to ensure that resources in
+that namespace maintained by other servers are deleted as well. We can do this
+using resource finalizers.
```

Comment:
Added more detail.

It will require extra logic on federated servers to listen for namespace delete events and then delete their resources (maybe in a separate controller?)
A rogue server can also prevent namespace deletion.

Feel free to suggest if you have some ideas. For now have added that we will improve this in a future proposal.